docs/getting_started/module.md

```diff
@@ -0,0 +1,259 @@
+12:53:50.455 metadata-server info    serving metadata on address [::]:40909              application: application
+12:53:50.455 kernel  info    stopped background module api-health-check          application: application
+12:53:50.455 kernel  info    stopped background module metadata-server           application: application
+12:53:50.455 kernel  info    stopped stage 1024                                  application: application
+12:53:50.455 kernel  info    stopping stage 0                                    application: application
+12:53:50.455 kernel  info    stopped stage 0                                     application: application
+12:53:50.455 kernel  info    leaving kernel                                      application: application
+```
+
+When running the code, _"foreground-module"_ gets to complete successfully, then as the kernel has no other foreground modules, _"hello-world"_ will have its `ctx` cancelled and thus log `Time to stop`.
+
+## Additional features:
+
+### Default values
+At runtime, every _Module_ object will have to implement the following interfaces:
+
+[embedmd]:# (../../pkg/kernel/module.go /type Module interface/ /\n}/)
+```go
+type Module interface {
+	// Run the module. If the provided context is canceled you have a
+	// few seconds (configurable with kernel.killTimeout) until your module
+	// is killed (via exit(1)). If you return from Run, it is assumed that
+	// your module is done executing and (depending on the type of your module)
+	// this might trigger a kernel shutdown. If you return an error, a kernel
+	// shutdown is also triggered.
+	Run(ctx context.Context) error
+}
+```
+
+This is the backbone of Gosoline module creation. Each _ModuleFactory_ is a function that returns one such _Module_.
+
+[embedmd]:# (../../pkg/kernel/module.go /type TypedModule interface/ /\n}/)
+```go
+type TypedModule interface {
+	IsEssential() bool
+	IsBackground() bool
+}
+```
+
+Internally, a kernel needs to know if a module is a background module, and if it is essential. To configure these properties you need to pass in _Options_ as parameters to `kernel.Add`. The default values, for a module without any specific options, is that this module is a non-essential foreground module. 
+
+[embedmd]:# (../../pkg/kernel/module.go /type StagedModule interface/ /\n}/)
+```go
+type StagedModule interface {
+	GetStage() int
+}
+```
+
+The stage mechanism is explained below.
+
+[embedmd]:# (../../pkg/kernel/module.go /type FullModule interface/ /\n}/)
+```go
+type FullModule interface {
+	Module
+	TypedModule
+	StagedModule
+}
+```
+
+A _FullModule_ needs to implement the other three interfaces.
+
+### Module and stages
+
+When a kernel starts, not all modules are started at the same time. They are grouped into stages, and all modules in a stage start at the same time. When a kernel starts, it begins with starting all modules in stage one, then those in stage two, and so on.
```

Comment:
This is misleading: We just spawn the modules, but they run whenever the go scheduler finds time. In your example above you can also see some module in stage 2048 running before a module in stage 1024. There is no order you can assume when starting modules, only an order when shutting them down

Reply:
This construct does exist to prevent a bug you could trigger by having a module do the equivalent of `func () error { return nil }`. Such a module could return before the second module of the stage was spawned, causing us to break the coffin we spawn the modules in. By waiting at least until all modules have been spawned, we can't spawn a module before the first (and only) module did already exit.

This does in no way at all ensure that modules of one stage run before modules of another stage, the scheduler could still select modules from stage 3 to run before stage 1